presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -605,6 +612,9 @@ private static String keyFromPath(Path path)
         if (key.endsWith(PATH_SEPARATOR)) {
             key = key.substring(0, key.length() - PATH_SEPARATOR.length());
         }
+        if (key.getBytes(UTF_8).length > MAX_S3_KEY_LENGTH) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Length of S3 key is longer than %d bytes: %s", MAX_S3_KEY_LENGTH, key));
```

Comment:
I'm concerned this violates the `FileSystem` interface by throwing a random non-`IOException`. Callers won't expect it and thus might not cleanup properly, etc. However, fixing it is more involved as the check must be done in multiple places (in some places like the retry loop, it should be moved outside loop).

Also, as a more general concern, we don't throw `PrestoException` or any `RuntimeException` in this code, and it seems best to keep it that way.

Thoughts?